Code/Framework/AzCore/AzCore/std/containers/span.h

```diff
@@ -0,0 +1,159 @@
+
+        using pointer = value_type*;
+        using const_pointer = const value_type*;
+
+        using reference = value_type&;
+        using const_reference = const value_type&;
+
+        using size_type = AZStd::size_t;
+        using difference_type = AZStd::ptrdiff_t;
+
+        using iterator = value_type*;
+        using const_iterator = const value_type*;
+        using reverse_iterator = AZStd::reverse_iterator<iterator>;
+        using const_reverse_iterator = AZStd::reverse_iterator<const_iterator>;
+
+        span()
+            : m_begin(nullptr)
+            , m_end(nullptr)
+        { }
+
+        ~span() = default;
+
+        span(const_pointer s, size_type length)
+            : m_begin(s)
+            , m_end(m_begin + length)
+        {
+            if (length == 0) erase();
+        }
+
+        span(const_pointer first, const_pointer last)
+            : m_begin(first)
+            , m_end(last)
+        { }
+
+        // We explicitly delete this constructor because it's too easy to accidentally 
+        // create a span to just the first element instead of an entire array.
+        span(const_pointer s) = delete;
+
+        template<AZStd::size_t N>
+        span(AZStd::array<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        span(AZStd::vector<value_type>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        template<AZStd::size_t N>
+        span(AZStd::fixed_vector<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        span(const span&) = default;
+
+        span(span&& other)
+            : span(other.m_begin, other.m_end)
+        {
+#if AZ_DEBUG_BUILD // Clearing the original pointers isn't necessary, but is good for debugging
+            other.m_begin = nullptr;
+            other.m_end = nullptr;
+#endif
+        }
+
+        span& operator=(const span& other) = default;
+
+        span& operator=(span&& other)
+        {
+            m_begin = other.m_begin;
+            m_end = other.m_end;
+#if AZ_DEBUG_BUILD // Clearing the original pointers isn't necessary, but is good for debugging
+            other.m_begin = nullptr;
+            other.m_end = nullptr;
+#endif
+            return *this;
+        }
+
+        size_type size() const { return m_end - m_begin; }
+
+        bool empty() const { return m_end == m_begin; }
+
+        pointer data() { return m_begin; }
+        const_pointer data() const { return m_begin; }
+
+        const_reference operator[](size_type index) const 
+        {
+            AZ_Assert(index < size(), "index value is out of range");
+            return m_begin[index]; 
+        }
+
+        void erase() { m_begin = m_end = nullptr; }
+
+        iterator         begin() const { return m_begin; }
+        iterator         end() const { return m_end; }
```

Comment:
I believe these should have const versions with const_iterator, and non-const versions with iterator.

Reply:
Added